supabase/functions/chat/index.ts

```diff
@@ -1,31 +1,60 @@
 import { serve } from "https://deno.land/std@0.170.0/http/server.ts";
 import OpenAI from "https://deno.land/x/openai@v4.26.0/mod.ts";
-
 import { corsHeaders } from "../common/cors.ts";
 import { supabaseClient } from "../common/supabaseClient.ts";
 import { ApplicationError, UserError } from "../common/errors.ts";
 
+
+interface ChatClient {
+  chat: {
+    completions: {
+      create: (params: { model: string; messages: Message[] }) => Promise<{ choices: Choice[] }>;
+    };
+  };
+  embeddings: {
+    create: (params: { model: string; input: string }) => Promise<{ data: any[] }>; // Adjust the return type according to your actual data structure
+  };
+}
+
+interface SearchResult {
+  id: number;
+  raw_text: string;
+  similarity: number;
+}
+
+interface Message {
+  role: string;
+  content: string;
+}
+
+interface Choice {
+  message: string;
+}
+
+// Current models available
+type ModelName = "nousresearch/nous-capybara-34b" | "mistral" | "gpt-4-0125-preview";
+
 async function generateResponse(
-  useOpenRouter,
-  useOllama,
-  openaiClient,
-  openRouterClient,
-  ollamaClient,
-  messages
+  useOpenRouter: boolean,
+  useOllama: boolean,
+  openaiClient: ChatClient,
+  openRouterClient: ChatClient | null,
+  ollamaClient: ChatClient | null,
+  messages: Message[]
```

Comment:
nitpick

These are probably not your changes, but since the clients seem to have the same api, we could just pass in one client here.
```suggestion
  client: ChatClient
  messages: Message[]
```